### ISSUE TITLE
Everything as Entity

### ISSUE THREAD

**Issue by valaphee** [valence-rs/valence#257]
### Is your feature request related to a problem? Please describe.

Following Bevy's Idiom of making everything an Entity and of my past experiences when it comes to ECS-driven Minecraft Servers, I was wondering if the concept of everything as Entity might also be a good idea for this project.

This would make a lot of logic more intuitive, queries can be more precise (and allow multi-threading chunks and worlds), also fits better with the Bevy idiom, and makes working with multiple worlds possible/easier.

### Describe the solution you'd like

This would build on top of partially fixed ids,

World Archetype:
- ChunkMap (ChunkX; ChunkZ -> Chunk Entity) LUT
- BlockMap (BlockKey -> Block Entity) LUT
- ItemMap (ItemKey -> Item Entity) LUT
- Weather, Time, etc.
- (weather_system, chunk_load, chunk_generation, ...)

Chunk Archetype:
- Terrain (Terrain and Biome Data)
- Parent (with childs of Actor Archetype)
- (block_update, block_break_queue_process, ...)

Block Archetype (Instanced):
- properties...

Chest Block Archetype (Specialized):
- properties...
- Inventory
- (hopper_ticking - not relevant here, ...)

Item Archetype (Instanced):
- properties...

Actor Archetype:
- properties...
- Location
- DamageQueue
- (broadcast_actor_move, broadcast_metadata_change, damage_queue_process, ...)

Player Archetype:
- properties...
- Location
- DamageQueue
- View (with view distance, which can be processed by the view system, which sends or request chunk loads/tickets)
- Inventory
- Connection
- ChunkLoader (keeps chunks loaded)
- (view_update, block_break_queue_update...)

### Additional context

For reference, my old prototype ECS-driven Server (for Bedrock Edition) https://github.com/valaphee/tesseract-kotlin/blob/master/src/main/kotlin/com/valaphee/tesseract/world/chunk/terrain/BlockUpdater.kt is using this system (worked well, despite written in Kotlin and using a way less powerful "ECS")

This system is also a bit inspired by Overwatch's ECS-driven Engine, and the Bedrock Edition ECS capabilities.

**Comment by valaphee** [comment]
I'm also already working on a prototype for this which can be found here: https://github.com/valaphee/tesseract/tree/main/crates/tesseract (called it tesseract as its also the same name as my Kotlin attempt a few years ago)

**Comment by dyc3** [comment]
duplicate of #199

**Comment by valaphee** [comment]
> duplicate of #199

Not really, as the focus is more on the world itself, and representing blocks, items, tile entities as entities.

**Comment by rj00a** [comment]
Having had time to experiment with this issue, I have decided to close.

"Everything is an entity" sounds great in theory, but in practice there are far too many usability and performance problems to make the approach work with bevy_ecs. In situations that are entirely random access lookups, ECS is usually the wrong tool for the job.

At the bare minimum, bevy_ecs needs [entity relations](https://ajmmertens.medium.com/building-games-in-ecs-with-entity-relationships-657275ba2c6c) to make queries more precise and to prevent invariants from being broken by users accidentally. We can definitely revisit once relations do eventually land, but even then I'm skeptical that it'll be enough for things like chunks, inventory items, and registry items.
